diff --git a/Release/CMakeLists.txt b/Release/CMakeLists.txt
index 13a3f01d..8f326366 100644
--- a/Release/CMakeLists.txt
+++ b/Release/CMakeLists.txt
@@ -166,7 +166,7 @@ elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
   message("-- Setting gcc options")
 
   set(WARNINGS -Wall -Wextra -Wunused-parameter -Wcast-align -Wcast-qual -Wconversion -Wformat=2 -Winit-self -Winvalid-pch -Wmissing-format-attribute -Wmissing-include-dirs -Wpacked -Wredundant-decls -Wunreachable-code)
-  set(LD_FLAGS "${LD_FLAGS} -Wl,-z,defs")
+  set(LD_FLAGS "${LD_FLAGS} -Wl,-z,defs,--exclude-libs,ALL")
 
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -fno-strict-aliasing")
   if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.8)
diff --git a/Release/cmake/cpprest_find_boost.cmake b/Release/cmake/cpprest_find_boost.cmake
index 3c857baf..d3f27bcd 100644
--- a/Release/cmake/cpprest_find_boost.cmake
+++ b/Release/cmake/cpprest_find_boost.cmake
@@ -46,7 +46,7 @@ function(cpprest_find_boost)
     endif()
     cpprestsdk_find_boost_android_package(Boost ${BOOST_VERSION} EXACT REQUIRED COMPONENTS random system thread filesystem chrono atomic)
   elseif(UNIX)
-    find_package(Boost REQUIRED COMPONENTS random system thread filesystem chrono atomic date_time regex)
+    find_package(Boost REQUIRED COMPONENTS log random system thread filesystem chrono atomic date_time regex)
   else()
     find_package(Boost REQUIRED COMPONENTS system date_time regex)
   endif()
@@ -86,6 +86,7 @@ function(cpprest_find_boost)
       )
     elseif(UNIX)
       target_link_libraries(cpprestsdk_boost_internal INTERFACE
+	Boost::log
         Boost::boost
         Boost::random
         Boost::system
@@ -96,6 +97,7 @@ function(cpprest_find_boost)
         Boost::date_time
         Boost::regex
       )
+      target_link_libraries(cpprestsdk_boost_internal PRIVATE "-Wl,--exclude-libs,ALL")
     else()
       target_link_libraries(cpprestsdk_boost_internal INTERFACE
         Boost::boost

