diff --git a/tensorflow/core/lib/io/cache.cc b/tensorflow/core/lib/io/cache.cc
index b5521b1752..feaa8d1b91 100644
--- a/tensorflow/core/lib/io/cache.cc
+++ b/tensorflow/core/lib/io/cache.cc
@@ -427,10 +427,10 @@ class ShardedLRUCache : public Cache {
     switch (limit - data) {
       case 3:
         h += static_cast<uint8_t>(data[2]) << 16;
-        ABSL_FALLTHROUGH_INTENDED;
+        [[fallthrough]];
       case 2:
         h += static_cast<uint8_t>(data[1]) << 8;
-        ABSL_FALLTHROUGH_INTENDED;
+        [[fallthrough]];
       case 1:
         h += static_cast<uint8_t>(data[0]);
         h *= m;
