diff --git a/tensorflow_serving/util/BUILD b/tensorflow_serving/util/BUILD
index 4fe1c1be..601e8675 100644
--- a/tensorflow_serving/util/BUILD
+++ b/tensorflow_serving/util/BUILD
@@ -4,7 +4,7 @@ load("//tensorflow_serving:serving.bzl", "serving_proto_library", "serving_proto
 
 package(
     default_visibility = [
-        "//tensorflow_serving:internal",
+        "//visibility:public",
     ],
     features = ["-layering_check"],
 )
diff --git a/tensorflow_serving/util/net_http/server/public/BUILD b/tensorflow_serving/util/net_http/server/public/BUILD
index e7f96d98..905ab4e4 100644
--- a/tensorflow_serving/util/net_http/server/public/BUILD
+++ b/tensorflow_serving/util/net_http/server/public/BUILD
@@ -2,9 +2,7 @@
 
 package(
     default_visibility = [
-        ":http_server_clients",
-        "//tensorflow_serving:internal",
-        "//tensorflow_serving/util/net_http:__subpackages__",
+        "//visibility:public"
     ],
 )

diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc b/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc
index a3b77b5..5e88ecf 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc
@@ -104,7 +104,10 @@ bool EvHTTPServer::Initialize() {
     NET_LOG(FATAL, "Failed to create evhttp.");
     return false;
   }
-
+  std::size_t maxBodySize = 1024 * 1024 * 1024;
+  evhttp_set_max_body_size(ev_http_, maxBodySize);
+  std::size_t maxHeadersSize = 8 * 1024;
+  evhttp_set_max_headers_size(ev_http_, maxHeadersSize);
   // By default libevents only allow GET, POST, HEAD, PUT, DELETE request
   // we have to manually turn OPTIONS flag on
   // documentation:
   // (http://www.wangafu.net/~nickm/libevent-2.0/doxygen/html/http_8h.html)
   evhttp_set_allowed_methods(
      ev_http_, EVHTTP_REQ_GET | EVHTTP_REQ_POST | EVHTTP_REQ_HEAD |
                    EVHTTP_REQ_PUT | EVHTTP_REQ_DELETE | EVHTTP_REQ_OPTIONS);
   evhttp_set_gencb(ev_http_, &DispatchEvRequestFn, this);

   return true;
