diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc b/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
index c8d0501b..401d8458 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
@@ -26,6 +26,8 @@ limitations under the License.
 #include <string>
 #include <vector>
 
+#include <event.h>
+
 #include "absl/strings/match.h"
 #include "absl/strings/string_view.h"
 #include "absl/types/span.h"
@@ -118,16 +120,51 @@ bool ParsedEvRequest::decode() {
   return true;
 }
 
+static void connection_close_callback(struct evhttp_connection *conn, void *arg) {
+  EvHTTPRequest* req = (EvHTTPRequest*)arg;
+  req->FlagDisconnected();
+  req->ExecuteDisconnectionCallback();
+}
+
 EvHTTPRequest::EvHTTPRequest(std::unique_ptr<ParsedEvRequest> request,
                              ServerSupport* server)
     : server_(server),
       parsed_request_(std::move(request)),
-      output_buf(nullptr) {}
+      output_buf(nullptr) {
+
+  struct evhttp_connection *conn = evhttp_request_get_connection(parsed_request_->request);
+  evhttp_connection_set_closecb(conn, connection_close_callback, (void*)this);
+
+  struct bufferevent *bev = evhttp_connection_get_bufferevent(conn);
+  bufferevent_enable(bev, EV_READ);
+}
 
 EvHTTPRequest::~EvHTTPRequest() {
+  NET_LOG(ERROR, "EvHTTPRequest::~EvHTTPRequest BEGIN");
+
+  struct evhttp_connection *conn = evhttp_request_get_connection(parsed_request_->request);
+  if (conn != NULL) {
+    evhttp_connection_set_closecb(conn, NULL, NULL);
+
+    // struct bufferevent *bev = evhttp_connection_get_bufferevent(conn);
+    // bufferevent_disable(bev, EV_READ);
+  }
+
+
   if (output_buf != nullptr) {
     evbuffer_free(output_buf);
   }
+  NET_LOG(ERROR, "EvHTTPRequest::~EvHTTPRequest END");
+}
+
+void EvHTTPRequest::RegisterDisconnectionCallback(std::function<void()> callback) { 
+  // TODO: Is data race possible? We should probably lock it
+  this->disconnected_callback_ = std::move(callback); 
+}
+
+void EvHTTPRequest::ExecuteDisconnectionCallback() {
+  if (this->disconnected_callback_)
+    this->disconnected_callback_(); 
 }
 
 absl::string_view EvHTTPRequest::uri_path() const {
@@ -342,8 +379,17 @@ void EvHTTPRequest::PartialReplyWithStatus(HTTPStatusCode status) {
   NET_LOG(FATAL, "PartialReplyWithStatus not implemented.");
 }
 
-void EvHTTPRequest::PartialReply() {
-  NET_LOG(FATAL, "PartialReplyWithStatus not implemented.");
+void EvHTTPRequest::PartialReply(std::string data) {
+  // TODO: Possibly avoid copy of data
+  bool result =
+      server_->EventLoopSchedule([this, data = std::move(data)]() mutable { EvPartialSendReply(std::move(data)); });
+
+  if (!result) {
+    NET_LOG(ERROR, "Failed to EventLoopSchedule PartialReply()");
+    Abort();
+    // TODO(wenboz): should have a forced abort that doesn't write back anything
+    // to the event-loop
+  }
 }
 
 ServerRequestInterface::CallbackStatus
@@ -371,6 +417,25 @@ void EvHTTPRequest::EvSendReply(HTTPStatusCode status) {
   delete this;
 }
 
+void EvHTTPRequest::EvPartialSendReply(std::string data) {
+  if (!this->is_reply_started_) {
+    evhttp_send_reply_start(parsed_request_->request, HTTP_OK, "reply start");
+    this->is_reply_started_ = true;
+  }
+  evbuffer_add(output_buf, data.data(), static_cast<int64_t>(data.size()));
+  evhttp_send_reply_chunk(parsed_request_->request, output_buf);
+}
+
+void EvHTTPRequest::EvPartialReplyEnd() {
+  if (!this->is_reply_started_) {
+    // Start before we end can end the reply
+    evhttp_send_reply_start(parsed_request_->request, HTTP_OK, "no messages");
+  }
+  evhttp_send_reply_end(parsed_request_->request);
+  server_->DecOps();
+  delete this;
+}
+
 void EvHTTPRequest::Reply() { ReplyWithStatus(HTTPStatusCode::OK); }
 
 // Treats this as 500 for now and let libevent decide what to do
@@ -381,6 +446,18 @@ void EvHTTPRequest::Abort() {
   delete this;
 }
 
+void EvHTTPRequest::PartialReplyEnd() {
+  bool result =
+      server_->EventLoopSchedule([this]() { EvPartialReplyEnd(); });
+
+  if (!result) {
+    NET_LOG(ERROR, "Failed to EventLoopSchedule PartialReplyEnd()");
+    Abort();
+    // TODO(wenboz): should have a forced abort that doesn't write back anything
+    // to the event-loop
+  }
+}
+
 }  // namespace net_http
 }  // namespace serving
 }  // namespace tensorflow
diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_request.h b/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
index 2f8e601d..2129c163 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
@@ -94,7 +94,7 @@ class EvHTTPRequest final : public ServerRequestInterface {
                             absl::string_view value) override;
 
   void PartialReplyWithStatus(HTTPStatusCode status) override;
-  void PartialReply() override;
+  void PartialReply(std::string data) override;
 
   CallbackStatus PartialReplyWithFlushCallback(
       std::function<void()> callback) override;
@@ -104,9 +104,16 @@ class EvHTTPRequest final : public ServerRequestInterface {
 
   void Abort() override;
 
+  void PartialReplyEnd() override;
+
   // Initializes the resource and returns false if any error.
   bool Initialize();
 
+  void FlagDisconnected() { this->is_disconnected_ = true; }
+  bool IsDisconnected() const override { return this->is_disconnected_; }
+  void RegisterDisconnectionCallback(std::function<void()> callback) override;
+  void ExecuteDisconnectionCallback();
+
   // Keeps a reference to the registered RequestHandlerOptions
   void SetHandlerOptions(const RequestHandlerOptions& handler_options) {
     this->handler_options_ = &handler_options;
@@ -114,6 +121,8 @@ class EvHTTPRequest final : public ServerRequestInterface {
 
  private:
   void EvSendReply(HTTPStatusCode status);
+  void EvPartialSendReply(std::string data);
+  void EvPartialReplyEnd();
 
   // Returns true if the data needs be uncompressed
   bool NeedUncompressGzipContent();
@@ -133,6 +142,11 @@ class EvHTTPRequest final : public ServerRequestInterface {
   std::unique_ptr<ParsedEvRequest> parsed_request_;
 
   evbuffer* output_buf;  // owned by this
+
+  bool is_reply_started_{false};
+  bool is_disconnected_{false};
+
+  std::function<void()> disconnected_callback_;
 };
 
 }  // namespace net_http
diff --git a/tensorflow_serving/util/net_http/server/public/server_request_interface.h b/tensorflow_serving/util/net_http/server/public/server_request_interface.h
index e5f4b05f..3205f0ab 100644
--- a/tensorflow_serving/util/net_http/server/public/server_request_interface.h
+++ b/tensorflow_serving/util/net_http/server/public/server_request_interface.h
@@ -144,7 +144,7 @@ class ServerRequestInterface {
   // PartialReply() is called is considered a programming error and
   // the underlying behavior is undefined.
   virtual void PartialReplyWithStatus(HTTPStatusCode status) = 0;
-  virtual void PartialReply() = 0;
+  virtual void PartialReply(std::string data) = 0;
 
   // Similar to PartialReply() but with an on_flush callback which will be
   // invoked when the response data has been completely flushed by the
@@ -182,6 +182,12 @@ class ServerRequestInterface {
   // by the server runtime.
   virtual void Abort() = 0;
 
+  virtual void PartialReplyEnd() = 0;
+
+  // Helpers for handling disconnection states
+  virtual bool IsDisconnected() const = 0;
+  virtual void RegisterDisconnectionCallback(std::function<void()> callback) = 0;
+
  protected:
   ServerRequestInterface() = default;
 
