diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc b/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
index c8d0501b..0484bf97 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
@@ -17,6 +17,8 @@ limitations under the License.
 
 #include "tensorflow_serving/util/net_http/server/internal/evhttp_request.h"
 
+#include <mutex>
+
 #include <zlib.h>
 
 #include <cassert>
@@ -26,6 +28,8 @@ limitations under the License.
 #include <string>
 #include <vector>
 
+#include <event.h>
+
 #include "absl/strings/match.h"
 #include "absl/strings/string_view.h"
 #include "absl/types/span.h"
@@ -118,18 +122,43 @@ bool ParsedEvRequest::decode() {
   return true;
 }
 
+static void connection_close_callback(struct evhttp_connection *conn, void *arg) {
+  EvHTTPRequest* req = (EvHTTPRequest*)arg;
+  req->ExecuteDisconnectionCallback();
+}
+
 EvHTTPRequest::EvHTTPRequest(std::unique_ptr<ParsedEvRequest> request,
                              ServerSupport* server)
     : server_(server),
       parsed_request_(std::move(request)),
-      output_buf(nullptr) {}
+      output_buf(nullptr) {
+  struct evhttp_connection *conn = evhttp_request_get_connection(parsed_request_->request);
+  evhttp_connection_set_closecb(conn, connection_close_callback, (void*)this);
+}
 
 EvHTTPRequest::~EvHTTPRequest() {
+  struct evhttp_connection *conn = evhttp_request_get_connection(parsed_request_->request);
+  if (conn != NULL) {
+    evhttp_connection_set_closecb(conn, NULL, NULL);
+  }
+
   if (output_buf != nullptr) {
     evbuffer_free(output_buf);
   }
 }
 
+void EvHTTPRequest::RegisterDisconnectionCallback(std::function<void()> callback) {
+  std::unique_lock<std::mutex> lk(this->disconnection_mx_);
+  this->disconnected_callback_ = std::move(callback);
+}
+
+void EvHTTPRequest::ExecuteDisconnectionCallback() {
+  std::unique_lock<std::mutex> lk(this->disconnection_mx_);
+  this->is_disconnected_ = true;
+  if (this->disconnected_callback_)
+    this->disconnected_callback_();
+}
+
 absl::string_view EvHTTPRequest::uri_path() const {
   return parsed_request_->path_and_query;
 }
@@ -338,12 +367,18 @@ void EvHTTPRequest::AppendResponseHeader(absl::string_view header,
   }
 }
 
-void EvHTTPRequest::PartialReplyWithStatus(HTTPStatusCode status) {
-  NET_LOG(FATAL, "PartialReplyWithStatus not implemented.");
+void EvHTTPRequest::PartialReplyWithStatus(std::string data, HTTPStatusCode status) {
+  bool result =
+      server_->EventLoopSchedule([this, data = std::move(data), status]() mutable { EvPartialSendReply(std::move(data), status); });
+
+  if (!result) {
+    NET_LOG(ERROR, "Failed to EventLoopSchedule PartialReply()");
+    Abort();
+  }
 }
 
-void EvHTTPRequest::PartialReply() {
-  NET_LOG(FATAL, "PartialReplyWithStatus not implemented.");
+void EvHTTPRequest::PartialReply(std::string data) {
+  return PartialReplyWithStatus(std::move(data), HTTPStatusCode::OK);
 }
 
 ServerRequestInterface::CallbackStatus
@@ -371,6 +406,33 @@ void EvHTTPRequest::EvSendReply(HTTPStatusCode status) {
   delete this;
 }
 
+void EvHTTPRequest::EvPartialSendReply(std::string data, HTTPStatusCode status) {
+  std::unique_lock<std::mutex> lk(this->disconnection_mx_);
+  if (this->is_disconnected_)
+    return;
+  if (!this->is_reply_started_) {
+    evhttp_send_reply_start(parsed_request_->request, static_cast<int>(status), "reply start");
+    this->is_reply_started_ = true;
+  }
+  evbuffer_add(output_buf, data.data(), static_cast<int64_t>(data.size()));
+  evhttp_send_reply_chunk(parsed_request_->request, output_buf);
+}
+
+void EvHTTPRequest::EvPartialReplyEnd() {
+  std::unique_lock<std::mutex> lk(this->disconnection_mx_);
+  if (!this->is_disconnected_) {
+    if (!this->is_reply_started_) {
+      // Start before we end can end the reply
+      evhttp_send_reply_start(parsed_request_->request, HTTP_OK, "no messages");
+    }
+    
+    evhttp_send_reply_end(parsed_request_->request);
+  }
+
+  server_->DecOps();
+  delete this;
+}
+
 void EvHTTPRequest::Reply() { ReplyWithStatus(HTTPStatusCode::OK); }
 
 // Treats this as 500 for now and let libevent decide what to do
@@ -381,6 +443,16 @@ void EvHTTPRequest::Abort() {
   delete this;
 }
 
+void EvHTTPRequest::PartialReplyEnd() {
+  bool result =
+      server_->EventLoopSchedule([this]() { EvPartialReplyEnd(); });
+
+  if (!result) {
+    NET_LOG(ERROR, "Failed to EventLoopSchedule PartialReplyEnd()");
+    Abort();
+  }
+}
+
 }  // namespace net_http
 }  // namespace serving
 }  // namespace tensorflow
diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_request.h b/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
index 2f8e601d..72e47bce 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
@@ -21,6 +21,7 @@ limitations under the License.
 #include <cstdint>
 #include <memory>
 #include <string>
+#include <mutex>
 
 #include "tensorflow_serving/util/net_http/server/internal/server_support.h"
 #include "tensorflow_serving/util/net_http/server/public/httpserver_interface.h"
@@ -93,8 +94,8 @@ class EvHTTPRequest final : public ServerRequestInterface {
   void AppendResponseHeader(absl::string_view header,
                             absl::string_view value) override;
 
-  void PartialReplyWithStatus(HTTPStatusCode status) override;
-  void PartialReply() override;
+  void PartialReplyWithStatus(std::string data, HTTPStatusCode status) override;
+  void PartialReply(std::string data) override;
 
   CallbackStatus PartialReplyWithFlushCallback(
       std::function<void()> callback) override;
@@ -104,9 +105,15 @@ class EvHTTPRequest final : public ServerRequestInterface {
 
   void Abort() override;
 
+  void PartialReplyEnd() override;
+
   // Initializes the resource and returns false if any error.
   bool Initialize();
 
+  bool IsDisconnected() const override { return this->is_disconnected_; }
+  void RegisterDisconnectionCallback(std::function<void()> callback) override;
+  void ExecuteDisconnectionCallback();
+
   // Keeps a reference to the registered RequestHandlerOptions
   void SetHandlerOptions(const RequestHandlerOptions& handler_options) {
     this->handler_options_ = &handler_options;
@@ -114,6 +121,8 @@ class EvHTTPRequest final : public ServerRequestInterface {
 
  private:
   void EvSendReply(HTTPStatusCode status);
+  void EvPartialSendReply(std::string data, HTTPStatusCode status);
+  void EvPartialReplyEnd();
 
   // Returns true if the data needs be uncompressed
   bool NeedUncompressGzipContent();
@@ -133,6 +142,12 @@ class EvHTTPRequest final : public ServerRequestInterface {
   std::unique_ptr<ParsedEvRequest> parsed_request_;
 
   evbuffer* output_buf;  // owned by this
+
+  bool is_reply_started_{false};
+  bool is_disconnected_{false};
+
+  std::function<void()> disconnected_callback_;
+  std::mutex disconnection_mx_;
 };
 
 }  // namespace net_http
diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc b/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc
index 36c925a8..20527907 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_server.cc
@@ -240,17 +240,17 @@ bool EvHTTPServer::StartAcceptingRequests() {
     ResolveEphemeralPort(ev_listener_, &port_);
   }
 
+  accepting_requests_.Notify();
+
   IncOps();
   server_options_->executor()->Schedule([this]() {
     NET_LOG(INFO, "Entering the event loop ...");
-    int result = event_base_dispatch(ev_base_);
-    NET_LOG(INFO, "event_base_dispatch() exits with value %d", result);
+    int result = event_base_loop(ev_base_, EVLOOP_NO_EXIT_ON_EMPTY);
+    NET_LOG(INFO, "event_base_loop() exits with value %d", result);
 
     DecOps();
   });
 
-  accepting_requests_.Notify();
-
   return true;
 }
 
diff --git a/tensorflow_serving/util/net_http/server/public/server_request_interface.h b/tensorflow_serving/util/net_http/server/public/server_request_interface.h
index e5f4b05f..40e87833 100644
--- a/tensorflow_serving/util/net_http/server/public/server_request_interface.h
+++ b/tensorflow_serving/util/net_http/server/public/server_request_interface.h
@@ -143,8 +143,8 @@ class ServerRequestInterface {
   // Trying to modify headers or specifying a status after the first
   // PartialReply() is called is considered a programming error and
   // the underlying behavior is undefined.
-  virtual void PartialReplyWithStatus(HTTPStatusCode status) = 0;
-  virtual void PartialReply() = 0;
+  virtual void PartialReplyWithStatus(std::string data, HTTPStatusCode status) = 0;
+  virtual void PartialReply(std::string data) = 0;
 
   // Similar to PartialReply() but with an on_flush callback which will be
   // invoked when the response data has been completely flushed by the
@@ -182,6 +182,12 @@ class ServerRequestInterface {
   // by the server runtime.
   virtual void Abort() = 0;
 
+  virtual void PartialReplyEnd() = 0;
+
+  // Helpers for handling disconnection states
+  virtual bool IsDisconnected() const = 0;
+  virtual void RegisterDisconnectionCallback(std::function<void()> callback) = 0;
+
  protected:
   ServerRequestInterface() = default;
 
