diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc b/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
index c8d0501b..f24a21d5 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_request.cc
@@ -342,8 +342,17 @@ void EvHTTPRequest::PartialReplyWithStatus(HTTPStatusCode status) {
   NET_LOG(FATAL, "PartialReplyWithStatus not implemented.");
 }
 
-void EvHTTPRequest::PartialReply() {
-  NET_LOG(FATAL, "PartialReplyWithStatus not implemented.");
+void EvHTTPRequest::PartialReply(std::string data) {
+  // TODO: Possibly avoid copy of data
+  bool result =
+      server_->EventLoopSchedule([this, data]() { EvPartialSendReply(data); });
+  
+  if (!result) {
+    NET_LOG(ERROR, "Failed to EventLoopSchedule PartialReply()");
+    Abort();
+    // TODO(wenboz): should have a forced abort that doesn't write back anything
+    // to the event-loop
+  }
 }
 
 ServerRequestInterface::CallbackStatus
@@ -371,6 +380,25 @@ void EvHTTPRequest::EvSendReply(HTTPStatusCode status) {
   delete this;
 }
 
+void EvHTTPRequest::EvPartialSendReply(std::string data) {
+  if (!this->is_reply_started_) {
+    evhttp_send_reply_start(parsed_request_->request, HTTP_OK, "reply start");
+    this->is_reply_started_ = true;
+  }
+  evbuffer_add(output_buf, data.data(), static_cast<int64_t>(data.size()));
+  evhttp_send_reply_chunk(parsed_request_->request, output_buf);
+}
+
+void EvHTTPRequest::EvPartialReplyEnd() {
+  if (!this->is_reply_started_) {
+    // Start before we end can end the reply
+    evhttp_send_reply_start(parsed_request_->request, HTTP_OK, "no messages");
+  }
+  evhttp_send_reply_end(parsed_request_->request);
+  server_->DecOps();
+  delete this;
+}
+
 void EvHTTPRequest::Reply() { ReplyWithStatus(HTTPStatusCode::OK); }
 
 // Treats this as 500 for now and let libevent decide what to do
@@ -381,6 +409,18 @@ void EvHTTPRequest::Abort() {
   delete this;
 }
 
+void EvHTTPRequest::PartialReplyEnd() {
+  bool result =
+      server_->EventLoopSchedule([this]() { EvPartialReplyEnd(); });
+
+  if (!result) {
+    NET_LOG(ERROR, "Failed to EventLoopSchedule PartialReplyEnd()");
+    Abort();
+    // TODO(wenboz): should have a forced abort that doesn't write back anything
+    // to the event-loop
+  }
+}
+
 }  // namespace net_http
 }  // namespace serving
 }  // namespace tensorflow
diff --git a/tensorflow_serving/util/net_http/server/internal/evhttp_request.h b/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
index 2f8e601d..ff51c570 100644
--- a/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
+++ b/tensorflow_serving/util/net_http/server/internal/evhttp_request.h
@@ -94,7 +94,7 @@ class EvHTTPRequest final : public ServerRequestInterface {
                             absl::string_view value) override;
 
   void PartialReplyWithStatus(HTTPStatusCode status) override;
-  void PartialReply() override;
+  void PartialReply(std::string data) override;
 
   CallbackStatus PartialReplyWithFlushCallback(
       std::function<void()> callback) override;
@@ -104,6 +104,8 @@ class EvHTTPRequest final : public ServerRequestInterface {
 
   void Abort() override;
 
+  void PartialReplyEnd() override;
+
   // Initializes the resource and returns false if any error.
   bool Initialize();
 
@@ -114,6 +116,8 @@ class EvHTTPRequest final : public ServerRequestInterface {
 
  private:
   void EvSendReply(HTTPStatusCode status);
+  void EvPartialSendReply(std::string data);
+  void EvPartialReplyEnd();
 
   // Returns true if the data needs be uncompressed
   bool NeedUncompressGzipContent();
@@ -133,6 +137,8 @@ class EvHTTPRequest final : public ServerRequestInterface {
   std::unique_ptr<ParsedEvRequest> parsed_request_;
 
   evbuffer* output_buf;  // owned by this
+
+  bool is_reply_started_{false};
 };
 
 }  // namespace net_http
diff --git a/tensorflow_serving/util/net_http/server/public/server_request_interface.h b/tensorflow_serving/util/net_http/server/public/server_request_interface.h
index e5f4b05f..7077a6c1 100644
--- a/tensorflow_serving/util/net_http/server/public/server_request_interface.h
+++ b/tensorflow_serving/util/net_http/server/public/server_request_interface.h
@@ -144,7 +144,7 @@ class ServerRequestInterface {
   // PartialReply() is called is considered a programming error and
   // the underlying behavior is undefined.
   virtual void PartialReplyWithStatus(HTTPStatusCode status) = 0;
-  virtual void PartialReply() = 0;
+  virtual void PartialReply(std::string data) = 0;
 
   // Similar to PartialReply() but with an on_flush callback which will be
   // invoked when the response data has been completely flushed by the
@@ -182,6 +182,8 @@ class ServerRequestInterface {
   // by the server runtime.
   virtual void Abort() = 0;
 
+  virtual void PartialReplyEnd() = 0;
+
  protected:
   ServerRequestInterface() = default;
 
