diff --git a/tensorflow/core/lib/io/cache.cc b/tensorflow/core/lib/io/cache.cc
index 4e5fa12474e..d12fd3a3d38 100644
--- a/tensorflow/core/lib/io/cache.cc
+++ b/tensorflow/core/lib/io/cache.cc
@@ -427,10 +427,10 @@ class ShardedLRUCache : public Cache {
     switch (limit - data) {
       case 3:
         h += static_cast<uint8_t>(data[2]) << 16;
-        ABSL_FALLTHROUGH_INTENDED;
+        [[fallthrough]];
       case 2:
         h += static_cast<uint8_t>(data[1]) << 8;
-        ABSL_FALLTHROUGH_INTENDED;
+        [[fallthrough]];
       case 1:
         h += static_cast<uint8_t>(data[0]);
         h *= m;
diff --git a/tensorflow/core/platform/default/BUILD b/tensorflow/core/platform/default/BUILD
index b633f98f268..ba9904a7d87 100644
--- a/tensorflow/core/platform/default/BUILD
+++ b/tensorflow/core/platform/default/BUILD
@@ -194,6 +194,18 @@ cc_library(
     ],
 )
 
+cc_library(
+    name = "log_macros",
+    hdrs = ["log_macros.h"],
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
+    name = "glog",
+    visibility = ["//visibility:public"],
+    deps = ["@com_github_glog_glog//:glog",],
+)
+
 cc_library(
     name = "logging",
     srcs = ["logging.cc"],
@@ -205,6 +217,7 @@ cc_library(
     ],
     textual_hdrs = ["logging.h"],
     deps = [
+        ":glog",
         "//tensorflow/core/platform",
         "//tensorflow/core/platform:env_time",
         "//tensorflow/core/platform:macros",
diff --git a/tensorflow/core/platform/default/logging.cc b/tensorflow/core/platform/default/logging.cc
index 337d01b3498..5f6e3358745 100644
--- a/tensorflow/core/platform/default/logging.cc
+++ b/tensorflow/core/platform/default/logging.cc
@@ -380,7 +380,6 @@ void LogString(const char* fname, int line, int severity,
   LogMessage(fname, line, severity) << message;
 }
 
-template <>
 void MakeCheckOpValueString(std::ostream* os, const char& v) {
   if (v >= 32 && v <= 126) {
     (*os) << "'" << v << "'";
@@ -389,7 +388,6 @@ void MakeCheckOpValueString(std::ostream* os, const char& v) {
   }
 }
 
-template <>
 void MakeCheckOpValueString(std::ostream* os, const signed char& v) {
   if (v >= 32 && v <= 126) {
     (*os) << "'" << v << "'";
@@ -398,7 +396,6 @@ void MakeCheckOpValueString(std::ostream* os, const signed char& v) {
   }
 }
 
-template <>
 void MakeCheckOpValueString(std::ostream* os, const unsigned char& v) {
   if (v >= 32 && v <= 126) {
     (*os) << "'" << v << "'";
@@ -408,7 +405,6 @@ void MakeCheckOpValueString(std::ostream* os, const unsigned char& v) {
 }
 
 #if LANG_CXX11
-template <>
 void MakeCheckOpValueString(std::ostream* os, const std::nullptr_t& v) {
   (*os) << "nullptr";
 }
diff --git a/tensorflow/core/platform/default/logging.h b/tensorflow/core/platform/default/logging.h
index dfee3a62c47..9b6b27b70a7 100644
--- a/tensorflow/core/platform/default/logging.h
+++ b/tensorflow/core/platform/default/logging.h
@@ -36,8 +36,13 @@ limitations under the License.
 #include "tensorflow/core/platform/macros.h"
 #include "tensorflow/core/platform/types.h"
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsign-compare" 
+#include "glog/logging.h"
+#pragma GCC diagnostic pop
 // TODO(mrry): Prevent this Windows.h #define from leaking out of our headers.
 #undef ERROR
+#define COMPACT_GOOGLE_LOG_QFATAL COMPACT_GOOGLE_LOG_ERROR
 
 namespace tensorflow {
 const int INFO = 0;            // base_logging::INFO;
@@ -103,54 +108,6 @@ class LogMessageNull : public std::basic_ostringstream<char> {
   ~LogMessageNull() override {}
 };
 
-#define _TF_LOG_INFO \
-  ::tensorflow::internal::LogMessage(__FILE__, __LINE__, ::tensorflow::INFO)
-#define _TF_LOG_WARNING \
-  ::tensorflow::internal::LogMessage(__FILE__, __LINE__, ::tensorflow::WARNING)
-#define _TF_LOG_ERROR \
-  ::tensorflow::internal::LogMessage(__FILE__, __LINE__, ::tensorflow::ERROR)
-#define _TF_LOG_FATAL \
-  ::tensorflow::internal::LogMessageFatal(__FILE__, __LINE__)
-
-#define _TF_LOG_QFATAL _TF_LOG_FATAL
-
-#define LOG(severity) _TF_LOG_##severity
-
-#ifdef IS_MOBILE_PLATFORM
-
-// Turn VLOG off when under mobile devices for considerations of binary size.
-#define VLOG_IS_ON(lvl) ((lvl) <= 0)
-
-#else
-
-// Otherwise, set TF_CPP_MAX_VLOG_LEVEL environment to update minimum log level
-// of VLOG, or TF_CPP_VMODULE to set the minimum log level for individual
-// translation units.
-#define VLOG_IS_ON(lvl)                                                     \
-  (([](int level, const char* fname) {                                      \
-    static const bool vmodule_activated =                                   \
-        ::tensorflow::internal::LogMessage::VmoduleActivated(fname, level); \
-    return vmodule_activated;                                               \
-  })(lvl, __FILE__))
-
-#endif
-
-#define VLOG(level)                                              \
-  TF_PREDICT_TRUE(!VLOG_IS_ON(level))                            \
-  ? (void)0                                                      \
-  : ::tensorflow::internal::Voidifier() &                        \
-          ::tensorflow::internal::LogMessage(__FILE__, __LINE__, \
-                                             tensorflow::INFO)
-
-// `DVLOG` behaves like `VLOG` in debug mode (i.e. `#ifndef NDEBUG`).
-// Otherwise, it compiles away and does nothing.
-#ifndef NDEBUG
-#define DVLOG VLOG
-#else
-#define DVLOG(verbose_level) \
-  while (false && (verbose_level) > 0) ::tensorflow::internal::LogMessageNull()
-#endif
-
 class LogEveryNState {
  public:
   bool ShouldLog(int n);
@@ -217,26 +174,6 @@ class LogEveryNSecState {
            logging_internal_stateful_condition_do_log;              \
            logging_internal_stateful_condition_do_log = false)
 
-// An instance of `LOG_EVERY_N` increments a hidden zero-initialized counter
-// every time execution passes through it and logs the specified message when
-// the counter's value is a multiple of `n`, doing nothing otherwise.  Each
-// instance has its own counter.  The counter's value can be logged by streaming
-// the symbol `COUNTER`.  `LOG_EVERY_N` is thread-safe.
-// Example:
-//
-//   for (const auto& user : all_users) {
-//     LOG_EVERY_N(INFO, 1000) << "Processing user #" << COUNTER;
-//     ProcessUser(user);
-//   }
-#define LOG_EVERY_N(severity, n)                       \
-  LOGGING_INTERNAL_STATEFUL_CONDITION(EveryN, true, n) \
-  LOG(severity)
-// `LOG_FIRST_N` behaves like `LOG_EVERY_N` except that the specified message is
-// logged when the counter's value is less than `n`.  `LOG_FIRST_N` is
-// thread-safe.
-#define LOG_FIRST_N(severity, n)                       \
-  LOGGING_INTERNAL_STATEFUL_CONDITION(FirstN, true, n) \
-  LOG(severity)
 // `LOG_EVERY_POW_2` behaves like `LOG_EVERY_N` except that the specified
 // message is logged when the counter's value is a power of 2.
 // `LOG_EVERY_POW_2` is thread-safe.
@@ -254,14 +191,6 @@ class LogEveryNSecState {
   LOGGING_INTERNAL_STATEFUL_CONDITION(EveryNSec, true, n_seconds) \
   LOG(severity)
 
-// CHECK dies with a fatal error if condition is not true.  It is *not*
-// controlled by NDEBUG, so the check will be executed regardless of
-// compilation mode.  Therefore, it is safe to do things like:
-//    CHECK(fp->Write(x) == 4)
-#define CHECK(condition)              \
-  if (TF_PREDICT_FALSE(!(condition))) \
-  LOG(FATAL) << "Check failed: " #condition " "
-
 // Function is overloaded for integral types to allow static const
 // integrals declared in classes and not defined to be used as arguments to
 // CHECK* macros. It's not encouraged though.
@@ -279,28 +208,6 @@ inline unsigned int GetReferenceableValue(unsigned int t) { return t; }
 inline int64_t GetReferenceableValue(int64_t t) { return t; }
 inline uint64 GetReferenceableValue(uint64 t) { return t; }
 
-// This formats a value for a failing CHECK_XX statement.  Ordinarily,
-// it uses the definition for operator<<, with a few special cases below.
-template <typename T>
-inline void MakeCheckOpValueString(std::ostream* os, const T& v) {
-  (*os) << v;
-}
-
-// Overrides for char types provide readable values for unprintable
-// characters.
-template <>
-void MakeCheckOpValueString(std::ostream* os, const char& v);
-template <>
-void MakeCheckOpValueString(std::ostream* os, const signed char& v);
-template <>
-void MakeCheckOpValueString(std::ostream* os, const unsigned char& v);
-
-#if LANG_CXX11
-// We need an explicit specialization for std::nullptr_t.
-template <>
-void MakeCheckOpValueString(std::ostream* os, const std::nullptr_t& v);
-#endif
-
 // A container for a string pointer which can be evaluated to a bool -
 // true iff the pointer is non-NULL.
 struct CheckOpString {
@@ -311,11 +218,6 @@ struct CheckOpString {
   string* str_;
 };
 
-// Build the error message string. Specify no inlining for code size.
-template <typename T1, typename T2>
-string* MakeCheckOpString(const T1& v1, const T2& v2,
-                          const char* exprtext) TF_ATTRIBUTE_NOINLINE;
-
 // A helper class for formatting "expr (V1 vs. V2)" in a CHECK_XX
 // statement.  See MakeCheckOpString for sample usage.  Other
 // approaches were considered: use of a template method (e.g.,
@@ -339,14 +241,6 @@ class CheckOpMessageBuilder {
   std::ostringstream* stream_;
 };
 
-template <typename T1, typename T2>
-string* MakeCheckOpString(const T1& v1, const T2& v2, const char* exprtext) {
-  CheckOpMessageBuilder comb(exprtext);
-  MakeCheckOpValueString(comb.ForVar1(), v1);
-  MakeCheckOpValueString(comb.ForVar2(), v2);
-  return comb.NewString();
-}
-
 // Helper functions for CHECK_OP macro.
 // We use the full name Check_EQ, Check_NE, etc. in case the file including
 // base/logging.h provides its own #defines for the simpler names EQ, NE, etc.
@@ -362,7 +256,7 @@ string* MakeCheckOpString(const T1& v1, const T2& v2, const char* exprtext) {
     if (TF_PREDICT_TRUE(v1 op v2))                                        \
       return NULL;                                                        \
     else                                                                  \
-      return ::tensorflow::internal::MakeCheckOpString(v1, v2, exprtext); \
+      return google::MakeCheckOpString(v1, v2, exprtext); \
   }                                                                       \
   inline string* name##Impl(int v1, int v2, const char* exprtext) {       \
     return name##Impl<int, int>(v1, v2, exprtext);                        \
@@ -377,7 +271,7 @@ TF_DEFINE_CHECK_OP_IMPL(Check_EQ, ==)
 
 inline string* Check_EQImpl(int v1, size_t v2, const char* exprtext) {
   if (TF_PREDICT_FALSE(v1 < 0))
-    ::tensorflow::internal::MakeCheckOpString(v1, v2, exprtext);
+    return google::MakeCheckOpString(v1, v2, exprtext);
 
   return Check_EQImpl(size_t(v1), v2, exprtext);
 }
@@ -408,7 +302,7 @@ inline string* Check_LEImpl(int v1, size_t v2, const char* exprtext) {
 
 inline string* Check_LEImpl(size_t v1, int v2, const char* exprtext) {
   if (TF_PREDICT_FALSE(v2 < 0))
-    return ::tensorflow::internal::MakeCheckOpString(v1, v2, exprtext);
+    return google::MakeCheckOpString(v1, v2, exprtext);
   return Check_LEImpl(v1, size_t(v2), exprtext);
 }
 
@@ -421,8 +315,7 @@ inline string* Check_LTImpl(int v1, size_t v2, const char* exprtext) {
 }
 
 inline string* Check_LTImpl(size_t v1, int v2, const char* exprtext) {
-  if (v2 < 0)
-    return ::tensorflow::internal::MakeCheckOpString(v1, v2, exprtext);
+    if (v2 < 0) return google::MakeCheckOpString(v1, v2, exprtext);
   return Check_LTImpl(v1, size_t(v2), exprtext);
 }
 
@@ -439,60 +332,6 @@ inline string* Check_GTImpl(const T1& v1, const T2& v2, const char* exprtext) {
 
 #undef TF_DEFINE_CHECK_OP_IMPL
 
-// In optimized mode, use CheckOpString to hint to compiler that
-// the while condition is unlikely.
-#define CHECK_OP_LOG(name, op, val1, val2)                     \
-  while (::tensorflow::internal::CheckOpString _result{        \
-      ::tensorflow::internal::name##Impl(                      \
-          ::tensorflow::internal::GetReferenceableValue(val1), \
-          ::tensorflow::internal::GetReferenceableValue(val2), \
-          #val1 " " #op " " #val2)})                           \
-  ::tensorflow::internal::LogMessageFatal(__FILE__, __LINE__) << *(_result.str_)
-
-#define CHECK_OP(name, op, val1, val2) CHECK_OP_LOG(name, op, val1, val2)
-
-// CHECK_EQ/NE/...
-#define CHECK_EQ(val1, val2) CHECK_OP(Check_EQ, ==, val1, val2)
-#define CHECK_NE(val1, val2) CHECK_OP(Check_NE, !=, val1, val2)
-#define CHECK_LE(val1, val2) CHECK_OP(Check_LE, <=, val1, val2)
-#define CHECK_LT(val1, val2) CHECK_OP(Check_LT, <, val1, val2)
-#define CHECK_GE(val1, val2) CHECK_OP(Check_GE, >=, val1, val2)
-#define CHECK_GT(val1, val2) CHECK_OP(Check_GT, >, val1, val2)
-#define CHECK_NOTNULL(val)                                 \
-  ::tensorflow::internal::CheckNotNull(__FILE__, __LINE__, \
-                                       "'" #val "' Must be non NULL", (val))
-
-#ifndef NDEBUG
-// DCHECK_EQ/NE/...
-#define DCHECK(condition) CHECK(condition)
-#define DCHECK_EQ(val1, val2) CHECK_EQ(val1, val2)
-#define DCHECK_NE(val1, val2) CHECK_NE(val1, val2)
-#define DCHECK_LE(val1, val2) CHECK_LE(val1, val2)
-#define DCHECK_LT(val1, val2) CHECK_LT(val1, val2)
-#define DCHECK_GE(val1, val2) CHECK_GE(val1, val2)
-#define DCHECK_GT(val1, val2) CHECK_GT(val1, val2)
-
-#else
-
-#define DCHECK(condition) \
-  while (false && (condition)) LOG(FATAL)
-
-// NDEBUG is defined, so DCHECK_EQ(x, y) and so on do nothing.
-// However, we still want the compiler to parse x and y, because
-// we don't want to lose potentially useful errors and warnings.
-// _DCHECK_NOP is a helper, and should not be used outside of this file.
-#define _TF_DCHECK_NOP(x, y) \
-  while (false && ((void)(x), (void)(y), 0)) LOG(FATAL)
-
-#define DCHECK_EQ(x, y) _TF_DCHECK_NOP(x, y)
-#define DCHECK_NE(x, y) _TF_DCHECK_NOP(x, y)
-#define DCHECK_LE(x, y) _TF_DCHECK_NOP(x, y)
-#define DCHECK_LT(x, y) _TF_DCHECK_NOP(x, y)
-#define DCHECK_GE(x, y) _TF_DCHECK_NOP(x, y)
-#define DCHECK_GT(x, y) _TF_DCHECK_NOP(x, y)
-
-#endif
-
 // These are for when you don't want a CHECK failure to print a verbose
 // stack trace.  The implementation of CHECK* in this file already doesn't.
 #define QCHECK(condition) CHECK(condition)
diff --git a/tensorflow/workspace2.bzl b/tensorflow/workspace2.bzl
index d46fab64b41..4af039c129b 100644
--- a/tensorflow/workspace2.bzl
+++ b/tensorflow/workspace2.bzl
@@ -564,10 +564,13 @@ def _tf_repositories():
     tf_http_archive(
         name = "zlib",
         build_file = "//third_party:zlib.BUILD",
-        sha256 = "91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9",
-        strip_prefix = "zlib-1.2.12",
+        sha256 = "b3a24de97a8fdbc835b9833169501030b8977031bcb54b3b3ac13740f846ab30",
+        strip_prefix = "zlib-1.2.13",
         system_build_file = "//third_party/systemlibs:zlib.BUILD",
-        urls = tf_mirror_urls("https://zlib.net/zlib-1.2.12.tar.gz"),
+        urls = [
+            "https://storage.googleapis.com/mirror.tensorflow.org/zlib.net/zlib-1.2.13.tar.gz",
+            "https://zlib.net/zlib-1.2.13.tar.gz",
+        ],
     )
 
     # LINT.IfChange
