diff --git a/tensorflow/lite/graph_info.h b/tensorflow/lite/graph_info.h
index 1093aa8bb72..07cb161adfb 100644
--- a/tensorflow/lite/graph_info.h
+++ b/tensorflow/lite/graph_info.h
@@ -122,7 +122,7 @@ using ControlEdges = std::vector<ControlEdge>;
 // (Example: with `greedily`, `control_edges.empty()`, and `nodes_to_partition
 // == {2, 3}`, the graph
 //
-//                    /------------\
+//                    /------------|
 //                    |            v
 // 0 --> 1 --> 2* --> 3*     4 --> 5
 //       |                   ^
@@ -134,7 +134,7 @@ using ControlEdges = std::vector<ControlEdge>;
 // With an additional control dependency `control_edges == {{3, 4}}` (notated
 // '==>'), execution of node 4 requires prior execution of node 3:
 //
-//                    /------------\
+//                    /------------|
 //                    |            v
 // 0 --> 1 --> 2* --> 3* ==> 4 --> 5
 //       |                   ^
