diff --git a/upb/upb.c b/upb/upb.c
index 266ea7d7..808d1f64 100644
--- a/upb/upb.c
+++ b/upb/upb.c
@@ -37,7 +37,10 @@ const char *upb_status_errmsg(const upb_status *status) { return status->msg; }
 void upb_status_seterrmsg(upb_status *status, const char *msg) {
   if (!status) return;
   status->ok = false;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
   strncpy(status->msg, msg, sizeof(status->msg));
+#pragma GCC diagnostic pop
   nullz(status);
 }
 
