--- a/mediapipe/framework/legacy_calculator_support.cc
+++ b/mediapipe/framework/legacy_calculator_support.cc
@@ -17,10 +17,16 @@
 namespace mediapipe {
 
 template <>
+#ifndef __APPLE__
+ABSL_CONST_INIT
+#endif
 thread_local CalculatorContext*
     LegacyCalculatorSupport::Scoped<CalculatorContext>::current_ = nullptr;
 template <>
+#ifndef __APPLE__
+ABSL_CONST_INIT
+#endif
 thread_local CalculatorContract*
     LegacyCalculatorSupport::Scoped<CalculatorContract>::current_ = nullptr;
 
 }  // namespace mediapipe
