diff --git a/third_party/model_api/model_api.bzl b/third_party/model_api/model_api.bzl
index 4f9f4a4..b1c30df 100644
--- a/third_party/model_api/model_api.bzl
+++ b/third_party/model_api/model_api.bzl
@@ -50,6 +50,6 @@ cmake(
     tags = ["requires-network"],
     deps = [
         "@mediapipe//mediapipe/framework/port:opencv_core",
-        "@mediapipe//third_party:openvino",
+        "@windows_openvino//:openvino_new_headers",
     ],
 )
@@ -59,6 +59,6 @@ cc_library(
     deps = [
         "@mediapipe//mediapipe/framework/port:opencv_core",
-        "@mediapipe//third_party:openvino",
+        "@windows_openvino//:openvino_new_headers",
         ":model_api_cmake",
     ],
     visibility = ["//visibility:public"],
diff --git a/third_party/BUILD b/third_party/BUILD
index 2f4f4de..3ab13d4 100644
--- a/third_party/BUILD
+++ b/third_party/BUILD
@@ -51,10 +51,10 @@ cc_library(
         "//mediapipe:macos": [
             "@com_github_glog_glog//:glog",
         ],
         "//mediapipe:windows": [
-            "@com_github_glog_glog_windows//:glog",
+            "@com_github_glog_glog//:glog",
         ],
         "//conditions:default": [
-            "@com_github_glog_glog//:glog",
+            "@com_github_glog_glog_no_gflags//:glog",
         ],
     }),
 )
